### PR TITLE
[Feature][3.4][Ready] Bulk action, bulk delete button

### PR DESCRIPTION
**Pull request by tabacitu**
Status: **100% working, needs feedback.**
Video: https://cl.ly/5f02ac5502b7

screen shot 2018-08-26 at 10 44 42

In order to enable this, you need to do:
```php
$this->crud->enableBulkActions();
$this->crud->addBulkDeleteButton();
```

This _particular button_ makes a separate AJAX call for each item you want to delete, so you don't need an additional method inside the Controller. But having a button that calls a method with "items to delete" as arguments is totally possible. Don't know whether to keep it this way or not - I personally like the fact that, if an entry fails, I get 4 success messages and one warning. But I'm not sure about this one. **Thoughts?**

To do:
- [x] refresh the DataTable after all AJAX calls have been made;
- [x] make checkboxes reappear checked when using pagination and filtering;
- [x] move texts to language file;
- [x] hide the ```custom_html``` and ```checkbox``` columns from the preview modal (on small viewports);
- [x] (probably) ```addBulkDeleteButton()``` that does the above;
- [x] (probably) checkbox on table headers that selects or deselects all visible checkboxes - but what happens with what's not visible / on other pagination pages???;
- [x] (maybe) allow users to select multiple items using SHIFT;
- [x] (maybe) one AJAX call instead of one for each entry;

### CODE REVIEW

**Comment by tswonke** [comment]
Great!

But sorry for asking: why do we need a "blank" column?

File: src/resources/views/list.blade.php [hunk 1]
```diff
@@ -43,8 +45,9 @@
                   <th
                     data-orderable="{{ var_export($column['orderable'], true) }}"
                     data-priority="{{ $column['priority'] }}"
+                    data-visible-in-modal="{{ (isset($column['visibleInModal']) && $column['visibleInModal'] == false) ? 'false' : 'true' }}"
                     >
-                    {{ $column['label'] }}
+                    {!! $column['label'] !!}
```
Comment (mikemand) [inline src/resources/views/list.blade.php]: I'm still new to using Backpack, but why was this changed? You can use `\Illuminate\Support\HtmlString` as the return value of `$column['label']` whenever you need HTML and keep using `{{ }}` for safe filtering of output. Same for line 62 below.

**Comment by tabacitu** [comment]
@tswonke great catch. It's the best solution I found for the checkbox in the table header to be perfectly aligned with the checkboxes in the rows - to be in their own separate column. 

It's the difference between this:

screen shot 2018-08-27 at 20 50 33

and this

screen shot 2018-08-27 at 20 54 08

If anybody has a better idea, I'm all ears. It's tricky because datatables AND datatables-responsive AND details-row all do stuff to the first column.

File: src/resources/views/list.blade.php (hunk 1 above)
Comment (tabacitu) [reply src/resources/views/list.blade.php]: It allows developers to put HTML inside the label. It's because we wanted general checkboxes in the table header, so we needed to be able to do this:
```php
$this->crud->addColumn([
            'type' => 'checkbox',
            'name' => 'bulk_actions',
            'label' => '<input type="checkbox" />',
        ]);
```

but now since we've opened the door, other developers could also use it for things like
```php
$this->crud->addColumn([
            'type' => 'select',
            'name' => 'tags',
            'label' => 'Tags <i class="fa fa-tags" data-toggle="tooltip" title="Tags that have been associated with this entry."></i>',
        ]);
```

I remember someone saying a while back that they needed this. At the time I disagreed too - I didn't see how/why you'd need to put HTML inside the label. So back then I totally agreed with you. But now, apparently, I have to track that person down to tell them they were right :-)

BUT.

The escaping is only changed for the ```label```, which is defined by the developer, never the user. Since it's something _the developer_ defines, I think the extra security is for nothing, but the extra flexibility is... for something :-))

If the developer wants to output HTML there, I say let him have it.

**Comment by tswonke** [comment]
ah, ok. thank you for the screens @tabacitu

File: src/resources/views/list.blade.php (hunk 1 above)
Comment (mikemand) [reply src/resources/views/list.blade.php]: It would be just as easy for the developer to do something like this instead:

```
$this->crud->addColumn([
    'type' => 'select',
    'name' => 'tags',
    'label' => new HtmlString('Tags <i class="fa fa-tags" data-toggle="tooltip" title="Tags that have been associated with this entry."></i>'),
]);
```

but if you want to keep it this way, that's fine. I was just wondering why the change. Thank you for answering. :)

**Comment by pxpm** [comment]
thumbs up for this.
 
how you think we can implement bulk actions not predefined ones ? 
For example, imagine i have a bulk action that is: Send email to all selected users! 

Should we be forcing the addBulkDeleteButton(); or we should use some magic here and allow the devs to define their own functions for bulks? 
like `$this->crud->addBulk('sendEmails','string_in_bulk_menu')` or `$this->crud->addBulk('bulkDelete','delete_selected_whatever') ` the first param could be the function name eg. and we can implement our own bulk actions knowing that we receive an array of id's for the current model ?

This is not about the delete, because that's something that could be "packaged" in backpack, but for custom bulk actions.

Thanks,
pxpm

**Comment by tabacitu** [comment]
@pxpm - you're totally right, Bulk Actions docs should include a "_how to create your own bulk action_" section - will do.

I don't think any special syntax is needed. Bulk actions use the existing ```bottom``` button stack, and this new ```checkbox``` column. So in order to create a bulk action, you only need to:
- ```$this->crud->enableBulkActions()``` - in order to show the checkbox column;
- create a new button and add it to your ```bottom``` stack with standard ```$this->crud->addButtonFromView('bla-bla')```;
- create a method in your EntityCrudController (or in a trait, if you want to re-use it for multiple CRUDs);
- create a route that points to the new method in you ```routes/backpack/custom.php``` or wherever else you hold your admin routes;

Come to think about it, I think the Clone operation could be a great example.
